src/server/nuclearnet/web_socket_proxy_nuclearnet_server.ts

```diff
@@ -93,6 +102,83 @@ class WebSocketServerClient {
   }
 
   private onPacket = (event: string, packet: NUClearNetPacket) => {
+    const peerKey = this.getPeer(this.processors.keys(), packet.peer)
+    if (peerKey) {
+      const processor = this.processors.get(peerKey)
+      if (processor) {
+        processor.onPacket(event, packet)
+      }
+    }
+  }
+
+  /**
+   * Look for the needle 'peer' inside the haystack 'peers'. They might not be the same object reference.
+   */
+  private getPeer(peers: Iterable<NUClearNetPeer>, peer: NUClearNetPeer): NUClearNetPeer | undefined {
+    return Array.from(peers).find(otherPeer => {
+      return otherPeer.name === peer.name && otherPeer.address === peer.address && otherPeer.port === peer.port
+    })
+  }
+}
+
+class RobotMessageProcessor {
+  private eventQueueSize: Map<string, number>
+
+  // The maximum number of packets of a unique type to send before receiving acknowledgements.
+  private limit: number
+
+  // The number of milliseconds before giving up on an acknowledge
+  private timeout: number
+
+  constructor(private socket: WebSocket,
+              private clock: Clock,
+              opts: { limit: number, timeout: number }) {
+    this.limit = opts.limit
+    this.timeout = opts.timeout
+    this.eventQueueSize = new Map()
+  }
+
+  public static of(socket: WebSocket) {
+    return new RobotMessageProcessor(socket, NodeSystemClock, { limit: 1, timeout: 5000 })
```

Comment:
Increasing `limit > 1` seems to have significant slowdown. I suspect a web worker pool would let us increase this 🙂